extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/results/QueryResultsFrameProcessor.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.msq.querykit.results;
+
+import it.unimi.dsi.fastutil.ints.IntSet;
+import org.apache.druid.frame.Frame;
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.frame.channel.WritableFrameChannel;
+import org.apache.druid.frame.processor.FrameProcessor;
+import org.apache.druid.frame.processor.FrameProcessors;
+import org.apache.druid.frame.processor.ReturnOrAwait;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+public class QueryResultsFrameProcessor implements FrameProcessor<Long>
+{
+  long numRows = 0L;
+  private final ReadableFrameChannel inChannel;
+  private final WritableFrameChannel outChannel;
+
+  public QueryResultsFrameProcessor(
+      final ReadableFrameChannel inChannel,
+      final WritableFrameChannel outChannel
+  )
+  {
+    this.inChannel = inChannel;
+    this.outChannel = outChannel;
+  }
+
+  @Override
+  public List<ReadableFrameChannel> inputChannels()
+  {
+    return Collections.singletonList(inChannel);
+  }
+
+  @Override
+  public List<WritableFrameChannel> outputChannels()
+  {
+    return Collections.singletonList(outChannel);
+  }
+
+  @Override
+  public ReturnOrAwait<Long> runIncrementally(final IntSet readableInputs) throws IOException
+  {
+    if (readableInputs.isEmpty()) {
+      return ReturnOrAwait.awaitAll(1);
+    }
+    if (inChannel.isFinished()) {
+      return ReturnOrAwait.returnObject(numRows);
+    }
+    writeFrame(inChannel.read());
+    return ReturnOrAwait.awaitAll(1);
+  }
+
+  @Override
+  public void cleanup() throws IOException
```

Comment:
This relates to the above point of opening the channel in the factory. 
1. It is fine if we open the channel in the factory and close it in the factory
2. Or if we open the channel in the processor and close it in the processor

However, currently we are opening the output channel in the processor and closing it in the factory. If the above point is addressed, then this cleanup method is fine. However currently, we should move it to the place where we are opening the output channel

Reply:
https://github.com/apache/druid/pull/14527#discussion_r1255240303